algoworld_contracts/swapper/swap_proxy.py

```diff
@@ -0,0 +1,119 @@
+"""
+MIT License
+
+Copyright (c) 2022 AlgoWorld
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+import dataclasses
+import sys
+
+from pyteal import (
+    Addr,
+    And,
+    Balance,
+    Bytes,
+    Cond,
+    Global,
+    Gtxn,
+    Int,
+    Mode,
+    Or,
+    Substring,
+    Txn,
+    TxnType,
+    compileTeal,
+)
+
+from algoworld_contracts.common.utils import parse_params
+
+"""
+Swapper Proxy Used for Storing Swap Configurations
+1. Activate Proxy
+2. Store Swap Configurations
+"""
+
+TEAL_VERSION = 6
+
+STORE_GSIZE = Int(2)
+STORE_FEE = 0
+STORE_PROXY_NOTE = 1
+
+
+@dataclasses.dataclass
+class SwapProxy:
+    swap_creator: str
+    version: str
+
+
+def swapper_proxy(cfg: SwapProxy):
+
+    is_proxy_store = And(
+        Global.group_size() == STORE_GSIZE,
+        Gtxn[STORE_FEE].type_enum() == TxnType.Payment,
+        Gtxn[STORE_PROXY_NOTE].type_enum() == TxnType.Payment,
+    )
+
+    return Cond([is_proxy_store, proxy_store(cfg)])
+
+
+def proxy_store(cfg: SwapProxy):
+    Balance(Txn.sender())
+
+    store_fee = And(
+        Gtxn[STORE_FEE].type_enum() == TxnType.Payment,
+        Or(
+            Gtxn[STORE_FEE].amount() == Int(110_000),
+            Gtxn[STORE_FEE].amount() == Int(10_000),
+        ),
+        Gtxn[STORE_FEE].sender() == Addr(cfg.swap_creator),
+        Gtxn[STORE_FEE].receiver() == Txn.sender(),
+        Gtxn[STORE_FEE].rekey_to() == Global.zero_address(),
+        Gtxn[STORE_FEE].close_remainder_to() == Global.zero_address(),
+    )
+
+    store_proxy_note = And(
+        Gtxn[STORE_PROXY_NOTE].type_enum() == TxnType.Payment,
+        Gtxn[STORE_PROXY_NOTE].amount() == Int(0),
+        Gtxn[STORE_PROXY_NOTE].sender() == Txn.sender(),
+        Gtxn[STORE_PROXY_NOTE].receiver() == Txn.sender(),
+        Gtxn[STORE_PROXY_NOTE].rekey_to() == Global.zero_address(),
+        Gtxn[STORE_PROXY_NOTE].close_remainder_to() == Global.zero_address(),
```

Comment:
Seems that `STORE_PROXY_NOTE` is an user zero Payment self-transaction. Since the transaction in not signed by the Smart Contract the check on `rekey_to` and `close_remainder_to` are not necessary. Am I understanding it correctly?

Reply:
Yes, that's correct. Wasn't aware that rekey and close_remainder are not necessary in this case. Could you expand on why it isn't needed?

Reply:
Both `close_remainder_to` and `rekey_to` take effect on who _signs_ the transaction, turning that into "dangerous" field when is actually the Smart Contract to sign and authorize the transaction. If, like in this case, the signer of the transaction is the user it would not be "dangerous" if the user wants to rekey or close its own account with this transaction, those are effects involving user's account (not the Smart Contract).